### DISCUSSION TITLE
Authenticating request to fetch local presets in self-hosting mode

### DISCUSSION THREAD

**Discussion by hisabimbola** [renovatebot/renovate#9638] [Request Help] [answered]
I'm trying to set up a local preset in a Github enterprise project. I have the base preset in a private repo in the organisation.

when I start renovate via github actions, it tries to resolve the preset with no authentication and this returns an error.

Is there anybody hosting a local preset in a private repo that authentication is needed? I think I need to somehow make renovate used the token provided in the config to make the request

Thanks for the help

**Comment by rarkins** [comment]
Is your GitHub Enterprise self-hosted?

And you are running via GitHub Actions?

Which token are you authenticating your action with? e.g. is it the default `GITHUB_TOKEN`, which is scoped to only allow access to the *current* repository it's running in?

**Reply by hisabimbola** [reply]
> Is your GitHub Enterprise self-hosted?

Yes

> And you are running via GitHub Actions?

Yes

> Which token are you authenticating your action with? e.g. is it the default GITHUB_TOKEN, which is scoped to only allow access to the current repository it's running in?

It's not the default github_token, but what I notice is that there is no token passed with the request at all

Below is the screenshot of the log
![image](https://user-images.githubusercontent.com/9654923/115350054-e39f8680-a1b4-11eb-8e45-54f8c3e8807b.png)

It seem, no authentication is added at all

**Reply by rarkins** [reply]
The token should normally be added to host rules after the platform is initialized, which happens early. Are you using a `local>` preset, and which version of Renovate?

**Reply by hisabimbola** [reply]
Yes, I'm trying to setup local presets and I'm using renovate:24.119. I just updated to latest to test.

What I see if that, it's trying to resolve the config on the line below
https://github.com/renovatebot/renovate/blob/8b8d7767bc48f5214b308a34ac6685b79ff7c10d/lib/config/index.ts#L56-L56

and at that point the repo is not yet initialised

**Reply by rarkins** [reply]
What's your action's config? Including how you reference the local preset.

**Reply by hisabimbola** [reply]
Here is the my config that is passed to RENOVATE_CONFIG_FILE env var

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "dryRun": true,
  "extends": ["local>company-org/renovate-presets:base"],
  "gitAuthor": "Renovate Bot <bot@renovateapp.com>",
  "onboarding": false,
  "platform": "github",
  "repositories": ["company-org/portal"],
  "includeForks": true,
  "endpoint": "https://company-url/api/v3",
  "docker": {
    "enabled": false
  },
  "ignoreNpmrcFile": true,
  "enabledManagers": ["npm"],
  "assigneesFromCodeOwners": true,
  "requireConfig": false,
  "printConfig": true
}
```

**Reply by rarkins** [reply]
And how do you pass the token, e.g. `RENOVATE_TOKEN`? And is it a PAT that you pass via secrets?

FYI I'm also adding #9644 which will result in debugging about hostRules addition and clearing, which would help

**Reply by hisabimbola** [reply]
Yes, via RENOVATE_TOKEN and yes, it's a PAT. I verified that the token has access to the repos

**Reply by rarkins** [reply]
It will take a while to flow though, but once that PR is in a release and the GitHub Action is updated with the new release, then you should get a log message like `DEBUG: Adding token authentication for api.github.com to hostRules` happening *before* the attempted preset resolution.

**Reply by hisabimbola** [reply]
Ok, I'll test and once the docker image is available

**Reply by hisabimbola** [reply]
@rarkins I've tested with the new docker image (24.119.4) and I don't see any log about hostrule log and I think that made sense because the repo is not even yet initialised at that stage where it's fetching the config

**Reply by rarkins** [reply]
The first time hostRules is set after the platform is initialised, not when the repository initialised. What's the new log?

**Reply by hisabimbola** [reply]
Yes, I meant the platform is not yet initialised. Nothing new in the logs
<img width="807" alt="Screenshot 2021-04-21 at 08 02 45" src="https://user-images.githubusercontent.com/9654923/115504346-18c0dd00-a278-11eb-82a9-62b6f11d963d.png">

You see from the screenshot all the logs.. Immediately we setup the code, next is the application trying to resolve the presets in the config

**Reply by rarkins** [reply]
Here's the log I get when trying to replicate your setup locally:

```
DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in /Users/rhys/src/renovatebot/renovate/config
DEBUG: Detected config in env RENOVATE_CONFIG
       "config": {"token": "***********", "extends": ["local>renovate-tests/preset3"]}
DEBUG: File config
       "config": {}
DEBUG: CLI config
       "config": {"repositories": ["renovate-tests/npm42"]}
DEBUG: Env config
       "config": {
         "token": "***********",
         "extends": ["local>renovate-tests/preset3"],
         "hostRules": []
       }
DEBUG: Combined config
       "config": {
         "token": "***********",
         "extends": ["local>renovate-tests/preset3"],
         "hostRules": [],
         "repositories": ["renovate-tests/npm42"]
       }
DEBUG: Using default github endpoint: https://api.github.com/
DEBUG: Authenticated as GitHub user: renovate-tests
DEBUG: Using platform gitAuthor: Renovate Testing <renovate-tests@redgc.com>
DEBUG: Adding token authentication for api.github.com to hostRules
DEBUG: Using baseDir: /var/folders/ts/vw100b2x2jjcwhhnrklb_l0r0000gn/T/renovate
DEBUG: Using cacheDir: /var/folders/ts/vw100b2x2jjcwhhnrklb_l0r0000gn/T/renovate/cache
DEBUG: Initializing Renovate internal cache into /var/folders/ts/vw100b2x2jjcwhhnrklb_l0r0000gn/T/renovate/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Clearing hostRules
DEBUG: Adding token authentication for api.github.com to hostRules
 INFO: Repository started (repository=renovate-tests/npm42)
```

That preset3 repo is private so can only work with auth.

If I change it to be a non-existent repo preset33 then it errors:

```
DEBUG: Commits limit = null
DEBUG: GitHub 404
       "url": "https://api.github.com/repos/renovate-tests/preset33/contents/default.json"
DEBUG: Failed to retrieve default.json from repo
       "statusCode": 404,
       "url": "https://api.github.com/repos/renovate-tests/preset33/contents/default.json"
DEBUG: default.json preset not found - trying renovate.json
DEBUG: GitHub 404
       "url": "https://api.github.com/repos/renovate-tests/preset33/contents/renovate.json"
DEBUG: Failed to retrieve renovate.json from repo
       "statusCode": 404,
       "url": "https://api.github.com/repos/renovate-tests/preset33/contents/renovate.json"
DEBUG: Preset fetch error
       "preset": "local>renovate-tests/preset33",
       "err": {
```

So I'm not sure why/how your app is taking that path of logic.

BTW in future please copy/paste logs in code blocks with backpacks and not screenshot as it's much more convenient for others.

**Reply by rarkins** [reply]
@viceice can you think if we're doing something different in the GitHub Action which could cause this?

**Reply by hisabimbola** [reply]
I was able to reproduce it [here](https://github.com/hisabimbola/github-renovate/actions).

Scenario 1:

Base preset Repo was public: [build successfully passes](https://github.com/hisabimbola/github-renovate/actions/runs/769958282)

Scenario 2:
Base preset repo is private: [build failed with 404](https://github.com/hisabimbola/github-renovate/actions/runs/769968543)

and the token is my personal token that has access to all my repos

**Reply by rarkins** [reply]
What if you remove the dry run? Also please see my comment below on removing the use of extends

**Reply by hisabimbola** [reply]
I removed the "dry-run" and it's still failing. Link [here](https://github.com/hisabimbola/github-renovate/runs/2398774708?check_suite_focus=true)

**Reply by hisabimbola** [reply]
I feel, this is properly never working for private repos. The authentication needs to either happen early or the preset are resolved at a later stage

**Reply by hisabimbola** [reply]
I created an issue https://github.com/renovatebot/renovate/issues/9665 to follow up on the fixing.

You can close if you think it's not a bug

**Reply by rarkins** [reply]
> I feel, this is properly never working for private repos. The authentication needs to either happen early or the preset are resolved at a later stage

There is most likely a problem, but I don't feel like you're paying attention to what I write or looking into the source code yourself instead of you don't. 

I have verified a private preset being resolved and shared the logs. If you keep insisting it's as simple as private presets not working at all then we've reached a dead end became the problem is more complex than that.

**Reply by viceice** [reply]
@hisabimbola Are you sure the renovate token can read `github>hisabimbola/renovate-base:base` preset? can you try to curl this file, full url can be copied from you logs.

**Reply by hisabimbola** [reply]
@viceice 

when I make a direct curl it works

```
curl --location --request GET 'https://api.github.com/repos/hisabimbola/renovate-base/contents/base.json' \
--header 'Accept: application/vnd.github.v3+json' \
--header 'Authorization: Bearer PAT_TOKEN'
```

Response

<details><summary>Response log</summary>

```
{
  "name": "base.json",
  "path": "base.json",
  "sha": "22a4dc1e033e060d100f9464bcdceb977c50e91c",
  "size": 161,
  "url": "https://api.github.com/repos/hisabimbola/renovate-base/contents/base.json?ref=main",
  "html_url": "https://github.com/hisabimbola/renovate-base/blob/main/base.json",
  "git_url": "https://api.github.com/repos/hisabimbola/renovate-base/git/blobs/22a4dc1e033e060d100f9464bcdceb977c50e91c",
  "download_url": "https://raw.githubusercontent.com/hisabimbola/renovate-base/main/base.json?token=ACJVFC4RIZOGWX7QGAAKBODAP73GS",
  "type": "file",
  "content": "ewogICIkc2NoZW1hIjogImh0dHBzOi8vZG9jcy5yZW5vdmF0ZWJvdC5jb20v\ncmVub3ZhdGUtc2NoZW1hLmpzb24iLAogICJkcnlSdW4iOiB0cnVlLAogICJl\neHRlbmRzIjogWyJjb25maWc6YmFzZSIsICJnaXRodWI+d2hpdGVzb3VyY2Uv\nbWVyZ2UtY29uZmlkZW5jZTpiZXRhIl0KfQo=\n",
  "encoding": "base64",
  "_links": {
    "self": "https://api.github.com/repos/hisabimbola/renovate-base/contents/base.json?ref=main",
    "git": "https://api.github.com/repos/hisabimbola/renovate-base/git/blobs/22a4dc1e033e060d100f9464bcdceb977c50e91c",
    "html": "https://github.com/hisabimbola/renovate-base/blob/main/base.json"
  }
}
```

</details>

**Reply by hisabimbola** [reply]
@rarkins I'm sorry if it seems I'm not paying attention. I'm trying to my best how to present the issue to you. Maybe it's working for some private repo but I'm not able to get it to work and why it doesn't work is why I'm trying to understand

Any help from you will be appreciated

**Reply by rarkins** [reply]
I wonder if this might start working after we release v25 thanks to this refactoring: https://github.com/renovatebot/renovate/pull/9394/commits/484604de21719f9ba8a5435bc20851ebbc292904

**Reply by hisabimbola** [reply]
Yes, I think that should fix it... is it possible to test it yet?

**Reply by viceice** [reply]
currently only from source, but hopefully released today 😅

**Reply by hisabimbola** [reply]
oh perfect... I'll be watching out for https://github.com/renovatebot/renovate/pull/9394 then

**Reply by hisabimbola** [reply]
Hello @rarkins @viceice 

I tested now with v25 and it worked. No 404 errors anymore. You can see https://github.com/hisabimbola/github-renovate/runs/2408476794

Thanks for fixing

**Comment by rarkins** [comment]
I suggest in the meantime putting your config in `RENOVATE_CONFIG` or a `config.js` file in the Action's repo and not using `extends` server-side like this.

**Comment by hisabimbola** [comment, marked as answer]
Issue is fixed in v25
